src/compositor/windowpixmapitem.cpp

```diff
@@ -542,131 +601,59 @@ void WindowPixmapItem::handleWindowSizeChanged()
 void WindowPixmapItem::itemDestroyed(QObject *)
 {
     m_item = 0;
-    if (!m_haveSnapshot) {
-        m_hasPixmap = false;
-        emit hasPixmapChanged();
+    m_surface.reset();
+    m_captureSurface.reset();
+    disconnect(window(), &QQuickWindow::beforeRendering, this, &WindowPixmapItem::captureSnapshot);
+    m_clearCapture = false;
+    update();
+    updateHasPixmap();
+}
+
+bool WindowPixmapItem::event(QEvent *event)
+{
+    if (event->type() == QEvent::User) {
+        m_clearNotified = false;
+        if (m_clearCapture) {
+            m_clearCapture = false;
+            m_captureSurface.reset();
+        }
+        updateHasPixmap();
+        return true;
+    } else {
+        return QQuickItem::event(event);
     }
 }
 
 QSGNode *WindowPixmapItem::updatePaintNode(QSGNode *oldNode, UpdatePaintNodeData *)
 {
-    SurfaceNode *node = static_cast<SurfaceNode *>(oldNode);
-
-    if (m_item == 0 && !m_haveSnapshot) {
-        if (node)
-            node->setTextureProvider(0, false);
-        delete node;
-        return 0;
+    if (m_captureSurface && !m_clearCapture) {
+        captureSnapshot();
```

Comment:
Probably pretty expensive, but it's only taken when the window surface is hidden and RenderWindow is not set or destroyed and RenderSnapshot is set.

m_captureSurface cleared after a capture and isn't set again until one of those events.   That said, calling captureSnaphot in updatePaintNode wouldn't have been necessary if it was connected to beforeSynchronizing instead beginRendering.